core/commonMain/src/implementations/immutableSet/TrieNode.kt

```diff
@@ -277,6 +326,194 @@ internal class TrieNode<E>(
+                            otherNodeCell as E
+                            thisCell.add(
+                                    otherNodeCell.hashCode(),
+                                    otherNodeCell,
+                                    shift + LOG_MAX_BRANCHING_FACTOR
+                            ).also {
+                                if (it === thisCell) intersectionSizeRef += 1
+                            }
+                        }
+                        // same as last case, but reversed
+                        otherIsNode -> @Suppress("UNCHECKED_CAST") {
+                            otherNodeCell as TrieNode<E>
+                            thisCell as E
+                            otherNodeCell.add(
+                                    thisCell.hashCode(),
+                                    thisCell,
+                                    shift + LOG_MAX_BRANCHING_FACTOR
+                            ).also {
+                                if (it === otherNodeCell) intersectionSizeRef += 1
+                            }
+                        }
+                        // both are just E => compare them
+                        thisCell == otherNodeCell -> thisCell.also { intersectionSizeRef += 1 }
+                        // both are just E, but different => make a collision-ish node
+                        else -> @Suppress("UNCHECKED_CAST") {
+                            thisCell as E
+                            otherNodeCell as E
+                            makeNode(
+                                    thisCell.hashCode(),
+                                    thisCell,
+                                    otherNodeCell.hashCode(),
+                                    otherNodeCell,
+                                    shift + LOG_MAX_BRANCHING_FACTOR,
+                                    null
+                            )
+                        }
+                    }
+                }
+            }
+        }
+        return when {
+            this.elementsEquals(newNode) -> this
+            otherNode.elementsEquals(newNode) -> otherNode
+            else -> newNode
+        }
+    }
+
+    fun mutableAddAll(otherNode: TrieNode<E>, shift: Int, mutator: PersistentHashSetBuilder<*>): TrieNode<E> {
+        if (this === otherNode) {
+            mutator.size -= this.calculateSize()
+            return this
+        }
+        if (shift > MAX_SHIFT) {
+            return mutableCollisionAddAll(otherNode, mutator)
+        }
+        // union mask contains all the bits from input masks
+        val newBitMap = bitmap or otherNode.bitmap
+        // first allocate the node and then fill it in
+        // we are doing a union, so all the array elements are guaranteed to exist
+        val mutableNode = when {
+            newBitMap == bitmap && ownedBy == mutator.ownership -> this
+            newBitMap == otherNode.bitmap && otherNode.ownedBy == mutator.ownership -> otherNode
```

Comment:
This case never occurs. Namely, `otherNode.ownedBy == mutator.ownership` to be true `otherNode` has to be a node in `mutator.node`, because ownership objects are unique to each builder. 
Thus, `otherNode.ownedBy == mutator.ownership` means `builder.addAll(builder)` has been called. So, `this === otherNode` and the method returns without reaching this point.
Anyway, modifying argument is conceptually not correct for `addAll` operation.

Reply:
Well, for me (conceptually) a union operation is more like `+`, being symmetrical and all, rather than having a receiver and an argument, so I try to treat both arguments equally.
While it makes sense for `+`, it does not make much of it for the builder variant, in that you are correct.